cpu/stm32l1/stmclk.c

```diff
@@ -0,0 +1,327 @@
+        clock_disable_clocks = RCC_CR_HSION | RCC_CR_MSION;
+    } else {
+        clock_cfgr_sw = RCC_CFGR_SW_HSI;
+        clock_cfgr_sw_rdy = RCC_CFGR_SWS_HSI;
+        clock_disable_clocks = RCC_CR_HSEON | RCC_CR_MSION;
+    }
+
+    if (CLOCK_USE_PLL) {
+        clock_cfgr_sw = RCC_CFGR_SW_PLL;
+        clock_cfgr_sw_rdy = RCC_CFGR_SWS_PLL;
+    }
+
+    tmpreg |= (uint32_t)clock_cfgr_sw;
+    RCC->CFGR = tmpreg;
+    while ((RCC->CFGR & (uint32_t)RCC_CFGR_SWS) != clock_cfgr_sw_rdy) {}
+    RCC->CR &= ~(clock_disable_clocks);
+#else
+    tmpreg |= (uint32_t)CLOCK_CFGR_SW;
+
+    RCC->CFGR = tmpreg;
+    /* Wait till clock is used as system clock source */
+    while ((RCC->CFGR & (uint32_t)RCC_CFGR_SWS) != CLOCK_CFGR_SW_RDY) {}
+
+    /* Disable other clock sources */
+    RCC->CR &= ~(CLOCK_DISABLE_OTHERS);
+#endif
+
+    cpu_clock_global = CLOCK_CORECLOCK;
+
+#if CLOCK_MSI
+    memcpy(cpu_clock_source, "MSI", 3);
+#elif defined(CLOCK_HS_MULTI)
+    uint32_t n = 0;
+    if (CLOCK_USE_PLL) {
+        memcpy(cpu_clock_source, "PLL", 3);
+        n += 3;
+    }
+    if (clock_source_rdy == RCC_CR_HSERDY) {
+        memcpy(cpu_clock_source + n, "/HSE", 4);
+    } else {
+        memcpy(cpu_clock_source + n, "/HSI", 4);
+    }
+#elif defined(CLOCK_HSI)
+    #if CLOCK_USE_PLL
+        memcpy(cpu_clock_source, "PLL/HSI", 7);
+    #elif
+        memcpy(cpu_clock_source, "HSI", 3);
+    #endif
+#elif defined(CLOCK_HSE)
+    #if CLOCK_USE_PLL
+        memcpy(cpu_clock_source, "PLL/HSE", 7);
+    #elif
+        memcpy(cpu_clock_source, "HSE", 3);
+    #endif
+#endif
+}
+
+void stmclk_switch_msi(uint32_t msi_range, uint32_t ahb_divider)
+{
+    RCC->CR |= RCC_CR_MSION;
+    while (!(RCC->CR & RCC_CR_MSIRDY)) {}
+
+    tmpreg = RCC->ICSCR;
+    tmpreg &= ~RCC_ICSCR_MSIRANGE;
+    tmpreg |= msi_range;
+    RCC->ICSCR = tmpreg;
+
+    tmpreg = RCC->CFGR;
+    tmpreg &= ~RCC_CFGR_HPRE;
+    tmpreg |= ahb_divider;
+    tmpreg &= ~RCC_CFGR_SW;
+    tmpreg |= RCC_CFGR_SW_MSI;
+    RCC->CFGR = tmpreg;
+
+    while ((RCC->CFGR & (uint32_t)RCC_CFGR_SWS) != RCC_CFGR_SWS_MSI) {}
+
+    if ((msi_range == RCC_ICSCR_MSIRANGE_1) || (msi_range == RCC_ICSCR_MSIRANGE_0)) {
+        /* Low-power run is only allowed at MSI Range 0 and 1 */
+        PWR->CR |= PWR_CR_LPSDSR | PWR_CR_LPRUN;
+    } else {
+        /* set Voltage Range 3 (1.2V) */
+        PWR->CR |= (PWR_CR_VOS_1 | PWR_CR_VOS_0);
+        while((PWR->CSR & PWR_CSR_VOSF) != 0) {}
+    }
+
+    /* Set latency = 0, disable prefetch and 64-bit access */
+    FLASH->ACR &= ~FLASH_ACR_LATENCY;
+    FLASH->ACR &= ~FLASH_ACR_PRFTEN;
+    FLASH->ACR &= ~FLASH_ACR_ACC64;
+    while (!(FLASH->SR & FLASH_SR_READY)) {}
+
+    /* Disable high speed clock sources and PLL */
+    tmpreg = RCC->CR;
+    tmpreg &= ~(RCC_CR_HSION | RCC_CR_HSEON);
+    tmpreg &= ~(RCC_CR_HSEBYP | RCC_CR_CSSON | RCC_CR_PLLON);
+    RCC->CR = tmpreg;
+
+    cpu_clock_global = 65536 * (1 << (msi_range >> 13));
```

Comment:
Can you use defines instead of only numbers here? And also add some comments about the usage of such constants.

Reply:
Going to remove this variable, it was used in timer.c to re-calculate it's frequency according to cpu_clock_global, but that should go in another PR if even.

Reply:
Thinking I should remove this. It isn't really needed unless runing MCU on MSI clock in run-mode. In that case CORE_CLOCK isn't the mcu actual speed and might need to recalculate timer, uart and other setting accordingly to this. But that would mean exposing a global variable and stuff like that. What do you think?